### PR TITLE
Thread tests

### PR DESCRIPTION
**Pull request by Bod9001**

### CODE REVIEW

File: Assets/Mirror/Runtime/NetworkConnectionToClient.cs [hunk 1]
```diff
@@ -44,18 +50,62 @@ public override void Disconnect()
...
         }
 
+        /// UNITYSTATION CODE /// logic for adding new dirty
+        public void AddDirty(NetworkIdentity Dirty)
+        {
+            if (EmptyIndex >= DirtyObserving.Length)
+            {
+                Debug.LogError($" Having to expand observer array expensive!!! how many do you have!!?!? {EmptyIndex} adding 1000");
+                Array.Resize(ref DirtyObserving, DirtyObserving.Length + 1000);
+            }
```
Comment (Alainx277) [inline Assets/Mirror/Runtime/NetworkConnectionToClient.cs:60]: Is there a reason to not use a `List` instead of an array? It can automatically resize, keeps track of the number of elements and can be pre-allocated by using `EnsureCapacity(size)`.

File: Assets/Mirror/Runtime/NetworkIdentity.cs [hunk 2]
```diff
@@ -41,10 +41,40 @@ public sealed class NetworkIdentity : MonoBehaviour
...
+                            }
+                        }
+                    }
+
+                    _isDirty = value;
+                }
+                else
+                {
+                    _isDirty = value;
+                }
```
Comment (Alainx277) [inline Assets/Mirror/Runtime/NetworkIdentity.cs:68]: Put `_isDirty = value;` once outside of the conditional.

File: Assets/Mirror/Runtime/NetworkIdentity.cs [hunk 3]
```diff
@@ -41,10 +41,40 @@ public sealed class NetworkIdentity : MonoBehaviour
...
         /// </summary>
-        [NonSerialized]
-        public bool isDirty;
+        public bool isDirty
+        {
+            set
+            {
+                if (_isDirty == false)
+                {
+                    if (value)
```
Comment (Alainx277) [inline Assets/Mirror/Runtime/NetworkIdentity.cs:52]: Combine these conditions.

File: Assets/Mirror/Runtime/NetworkServer.cs [hunk 4]
```diff
@@ -1637,26 +1640,33 @@ static void BroadcastToConnection(NetworkConnectionToClient connection)
...
-                    if (serialization != null)
+                    if (identity == null)
                     {
-                        EntityStateMessage message = new EntityStateMessage
-                        {
-                            netId = identity.netId,
-                            payload = serialization.ToArraySegment()
-                        };
-                        connection.Send(message);
+                        Debug.LogError("oh no!");
```
Comment (Alainx277) [inline Assets/Mirror/Runtime/NetworkServer.cs:1652]: There should be a proper error message (and probably a `continue`).

File: Assets/Mirror/Runtime/NetworkWriterPool.cs [hunk 5]
```diff
@@ -14,35 +15,144 @@ public sealed class PooledNetworkWriter : NetworkWriter, IDisposable
     /// <summary>Pool of NetworkWriters to avoid allocations.</summary>
     public static class NetworkWriterPool
     {
+        /// UNITYSTATION CODE /// multiple pools to reduce locks waiting
```
Comment (Alainx277) [inline Assets/Mirror/Runtime/NetworkWriterPool.cs:18]: 1. How would checking multiple booleans and then locking reduce lock contention if the lock is immediately released after taking from the pool?
2. Why did you remove the thread local?

**Review by Alainx277 — CHANGES_REQUESTED** [review]
Logic seems okay. There's just a lot of overcomplication in my opinion, I can accept those if we have benchmarks that show there's a clear benefit.

File: Assets/Mirror/Runtime/NetworkWriterPool.cs (hunk 5 above)
Comment (Bod9001) [reply Assets/Mirror/Runtime/NetworkWriterPool.cs:18]: 2 problem is called network writers would go across threads, meaning it would just dump on one thread and the rest would end up having to make new ones

Comment (Bod9001) [reply Assets/Mirror/Runtime/NetworkWriterPool.cs:18]: you have multiple pools to take from, so therefore you don't have to wait for the other one to finish taking, you can just take from another one

File: Assets/Mirror/Runtime/NetworkServer.cs (hunk 4 above)
Comment (Bod9001) [reply Assets/Mirror/Runtime/NetworkServer.cs:1652]: oh yeah am I will remove that since it takes performance up, was debug

File: Assets/Mirror/Runtime/NetworkConnectionToClient.cs (hunk 1 above)
Comment (Bod9001) [reply Assets/Mirror/Runtime/NetworkConnectionToClient.cs:60]: I think array is a tiny bit faster?

Comment (Alainx277) [reply Assets/Mirror/Runtime/NetworkConnectionToClient.cs:60]: Strongly doubt it, please benchmark it or get rid of all the array management code.

File: Assets/Mirror/Runtime/NetworkWriterPool.cs (hunk 5 above)
Comment (Alainx277) [reply Assets/Mirror/Runtime/NetworkWriterPool.cs:18]: > 2 problem is called network writers would go across threads, meaning it would just dump on one thread and the rest would end up having to make new ones

I'm sorry but I can't understand that sentence. Could you explain it again?

Comment (Alainx277) [reply Assets/Mirror/Runtime/NetworkWriterPool.cs:18]: > you have multiple pools to take from, so therefore you don't have to wait for the other one to finish taking, you can just take from another one

That's what the code does. I doubt it's really faster with the overhead of checking booleans and checking what pool to return it to. Please benchmark.
